File: sdks/python/src/untrace/metrics.py
"""Metrics implementation for the Untrace SDK."""

import logging
from typing import Any, Dict, Optional
from opentelemetry import metrics
from opentelemetry.metrics import Counter, Histogram, Meter

from .types import TokenUsage, Cost


logger = logging.getLogger(__name__)


class UntraceMetrics:
    """Metrics collection for Untrace SDK."""

    def __init__(self, meter_name: str = "untrace-sdk"):
        """Initialize the metrics collector.

        Args:
            meter_name: Name of the meter
        """
        self._meter: Meter = metrics.get_meter(meter_name)

        # Create metrics instruments
        self._token_counter = self._meter.create_counter(
            name="untrace.tokens.total",
            description="Total number of tokens processed",
            unit="1"
        )

        self._latency_histogram = self._meter.create_histogram(
            name="untrace.latency",
            description="Latency of operations",
            unit="ms"
        )

        self._error_counter = self._meter.create_counter(
            name="untrace.errors.total",
            description="Total number of errors",
            unit="1"
        )

        self._cost_counter = self._meter.create_counter(
            name="untrace.cost.total",
            description="Total cost of operations",
            unit="USD"
        )

    def record_token_usage(self, usage: TokenUsage) -> None:
        """Record token usage metrics.

        Args:
            usage: Token usage information
        """
        try:
            attributes = {
                "model": usage.model,
                "provider": usage.provider,
            }

            if usage.prompt_tokens:
                self._token_counter.add(
                    usage.prompt_tokens,
                    {**attributes, "token_type": "prompt"}
                )

            if usage.completion_tokens:
                self._token_counter.add(
                    usage.completion_tokens,
                    {**attributes, "token_type": "completion"}
                )

            if usage.total_tokens:
                self._token_counter.add(
                    usage.total_tokens,
                    {**attributes, "token_type": "total"}
                )

        except Exception as error:
            logger.error("Failed to record token usage: %s", error)

    def record_latency(self, duration_ms: float, attributes: Optional[Dict[str, Any]] = None) -> None:
        """Record latency metrics.

        Args:
            duration_ms: Duration in milliseconds
            attributes: Additional attributes
        """
        try:
            attrs = attributes or {}
            self._latency_histogram.record(duration_ms, attrs)
        except Exception as error:
            logger.error("Failed to record latency: %s", error)

    def record_error(self, error: Exception, attributes: Optional[Dict[str, Any]] = None) -> None:
        """Record error metrics.

        Args:
            error: Exception that occurred
            attributes: Additional attributes
        """
        try:
            attrs = {
                "error_type": type(error).__name__,
                "error_message": str(error),
                **(attributes or {})
            }
            self._error_counter.add(1, attrs)
        except Exception as error:
            logger.error("Failed to record error: %s", error)

    def record_cost(self, cost: Cost, attributes: Optional[Dict[str, Any]] = None) -> None:
        """Record cost metrics.

        Args:
            cost: Cost information
            attributes: Additional attributes
        """
        try:
            attrs = {
                "model": cost.model,
                "provider": cost.provider,
                "currency": cost.currency,
                **(attributes or {})
            }

            if cost.prompt:
                self._cost_counter.add(
                    cost.prompt,
                    {**attrs, "cost_type": "prompt"}
                )

            if cost.completion:
                self._cost_counter.add(
                    cost.completion,
                    {**attrs, "cost_type": "completion"}
                )

            self._cost_counter.add(
                cost.total,
                {**attrs, "cost_type": "total"}
            )

        except Exception as error:
            logger.error("Failed to record cost: %s", error)

    def get_meter(self) -> Meter:
        """Get the underlying meter.

        Returns:
            OpenTelemetry meter
        """
        return self._meter

File: sdks/go/untrace/metrics.go
package untrace

import (
	"context"
	"fmt"
	"time"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/metric"
)

// untraceMetrics implements the Metrics interface
type untraceMetrics struct {
	meter metric.Meter
}

// NewMetrics creates a new Untrace metrics instance
func NewMetrics(meter metric.Meter) Metrics {
	return &untraceMetrics{
		meter: meter,
	}
}

// RecordTokenUsage records token usage metrics
func (m *untraceMetrics) RecordTokenUsage(usage TokenUsage) {
	attrs := []attribute.KeyValue{
		attribute.String("model", usage.Model),
		attribute.String("provider", usage.Provider),
	}

	// Create counters for different token types
	promptCounter, _ := m.meter.Int64Counter("llm.prompt.tokens")
	completionCounter, _ := m.meter.Int64Counter("llm.completion.tokens")
	totalCounter, _ := m.meter.Int64Counter("llm.total.tokens")

	if usage.PromptTokens > 0 {
		promptCounter.Add(context.Background(), int64(usage.PromptTokens), metric.WithAttributes(attrs...))
	}
	if usage.CompletionTokens > 0 {
		completionCounter.Add(context.Background(), int64(usage.CompletionTokens), metric.WithAttributes(attrs...))
	}
	if usage.TotalTokens > 0 {
		totalCounter.Add(context.Background(), int64(usage.TotalTokens), metric.WithAttributes(attrs...))
	}
}

// RecordLatency records latency metrics
func (m *untraceMetrics) RecordLatency(duration time.Duration, attributes map[string]interface{}) {
	attrs := m.buildAttributes(attributes)

	histogram, _ := m.meter.Float64Histogram("llm.latency")
	histogram.Record(context.Background(), duration.Seconds(), metric.WithAttributes(attrs...))
}

// RecordError records error metrics
func (m *untraceMetrics) RecordError(err error, attributes map[string]interface{}) {
	attrs := m.buildAttributes(attributes)
	attrs = append(attrs, attribute.String("error.type", err.Error()))

	counter, _ := m.meter.Int64Counter("llm.errors")
	counter.Add(context.Background(), 1, metric.WithAttributes(attrs...))
}

// RecordCost records cost metrics
func (m *untraceMetrics) RecordCost(cost Cost) {
	attrs := []attribute.KeyValue{
		attribute.String("model", cost.Model),
		attribute.String("provider", cost.Provider),
		attribute.String("currency", cost.Currency),
	}

	// Create counters for different cost components
	promptCounter, _ := m.meter.Float64Counter("llm.cost.prompt")
	completionCounter, _ := m.meter.Float64Counter("llm.cost.completion")
	totalCounter, _ := m.meter.Float64Counter("llm.cost.total")

	if cost.Prompt > 0 {
		promptCounter.Add(context.Background(), cost.Prompt, metric.WithAttributes(attrs...))
	}
	if cost.Completion > 0 {
		completionCounter.Add(context.Background(), cost.Completion, metric.WithAttributes(attrs...))
	}
	if cost.Total > 0 {
		totalCounter.Add(context.Background(), cost.Total, metric.WithAttributes(attrs...))
	}
}

// buildAttributes converts a map of attributes to OpenTelemetry attributes
func (m *untraceMetrics) buildAttributes(attrs map[string]interface{}) []attribute.KeyValue {
	var result []attribute.KeyValue

	for key, value := range attrs {
		switch v := value.(type) {
		case string:
			result = append(result, attribute.String(key, v))
		case int:
			result = append(result, attribute.Int(key, v))
		case int64:
			result = append(result, attribute.Int64(key, v))
		case float64:
			result = append(result, attribute.Float64(key, v))
		case bool:
			result = append(result, attribute.Bool(key, v))
		case []string:
			result = append(result, attribute.StringSlice(key, v))
		case []int:
			result = append(result, attribute.IntSlice(key, v))
		case []float64:
			result = append(result, attribute.Float64Slice(key, v))
		default:
			// Convert to string as fallback
			result = append(result, attribute.String(key, fmt.Sprintf("%v", v)))
		}
	}

	return result
}
